Top CWV recommendations blog post (#9374)
GoogleChrome/web.dev: src/site/content/en/blog/top-cwv-2023/index.md

- Recommendations that are **relevant and applicable to most sites**
- Recommendations that are **realistic for most developers to implement**

Over the past year we've spent a lot of time auditing the full set of performance recommendations we make, and assessing each of them (both qualitatively and quantitatively) against the above three criteria.

This post outlines our top recommendations to improve performance for each of the [Core Web Vitals](/vitals/#core-web-vitals) metrics. If you're new to web performance, or if you're trying to decide what will give you the biggest bang for your buck, we think these recommendations are the best place to start.

## Largest Contentful Paint (LCP)

Our first set of recommendations are for [Largest Contentful Paint (LCP)](/lcp/), which is a measure of load performance. Of the three Core Web Vitals metrics, LCP is the one that the largest number of sites struggle with—only [about half](https://datastudio.google.com/s/nw4gcbKA5o4) of all sites on the web today meet the [recommended threshold](/lcp/#what-is-a-good-lcp-score)—so let's start there.

### Ensure the LCP resource is discoverable from the HTML source

> According to the [2022 Web Almanac](https://almanac.httparchive.org/en/2022/) by HTTP Archive, [72%](https://almanac.httparchive.org/en/2022/performance#fig-8) of mobile pages have an image as their LCP element, which means that for most sites to ensure their LCP is good, they'll need to ensure those images can load quickly.

Author: Done

> According to the [2022 Web Almanac](https://almanac.httparchive.org/en/2022/) by HTTP Archive, [72%](https://almanac.httparchive.org/en/2022/performance#fig-8) of mobile pages have an image as their LCP element, which means that for most sites to ensure their LCP is good, they'll need to ensure those images can load quickly.

Reviewer: Good catch. How about this?

```suggestion
According to the [2022 Web Almanac](https://almanac.httparchive.org/en/2022/) by HTTP Archive, [72%](https://almanac.httparchive.org/en/2022/performance#fig-8) of mobile pages have an image as their LCP element, which means that for most sites to optimize their LCP, they'll need to ensure those images can load quickly.
```

Another critical aspect of prioritizing the LCP resource is to ensure you don't do anything that causes it to be **deprioritized**, such as adding the `loading="lazy"` attribute. Today, [10%](https://almanac.httparchive.org/en/2022/performance#lcp-lazy-loading) of pages actually set `loading="lazy"` on their LCP image. Beware of image optimization solutions that indiscriminately apply lazy-loading behavior to all images. If they provide a way to override that behavior, be sure to use it for the LCP image. If you're not sure which image will be the LCP, try using heuristics to pick a reasonable candidate.

Deferring non-critical resources is another way to effectively boost the relative priority of the LCP resource. For example, scripts that are not powering the user interface (like analytics scripts or social widgets) can be safely postponed until after the `load` event fires, which ensures they won't compete with other critical resources (such as the LCP resource) for network bandwidth.

To summarize, you should follow these best practices to ensure that the LCP resource is loaded early, and at high priority:

> - **Add `fetchpriority="high"` to the `<img>` tag of your LCP image.** If the LCP resource is loaded via a` <link rel="preload">` tag, fear not because you can also set `fetchpriority="high"` on that!\
> - **Never set `loading="lazy"` on the `<img>` tag of your LCP image.** Doing this will deprioritize your image and delay when it starts loading.\

Reviewer 2: ```suggestion
- **Add `fetchpriority="high"` to the `<img>` tag of your LCP image.** If the LCP resource is loaded via a` <link rel="preload">` tag, fear not because you can also set `fetchpriority="high"` on that!
- **Never set `loading="lazy"` on the `<img>` tag of your LCP image.** Doing this will deprioritize your image and delay when it starts loading.
```

Author: Done

> - **Defer non-critical resources when possible.** Either by moving them to the end of your document, or loading them asynchronously via script.

Reviewer 2: ```suggestion
- **Defer non-critical resources when possible.** Either by moving them to the end of your document, using native lazy loading for [images](https://web.dev/browser-level-image-lazy-loading/) or [iframes](https://web.dev/iframe-lazy-loading/), or loading them asynchronously via JavaScript.
```

Author: Done

### Use a CDN to optimize document and resource TTFB

The previous two recommendations focused on making sure your LCP resource is discovered early and prioritized so it can start loading right away. The final piece to this puzzle is making sure the initial document response arrives as quickly as possible too.

The browser cannot start loading any subresources until it receives the first byte of the initial HTML document response, and the sooner that happens, the sooner everything else can start happening as well.

This time is known as [Time to First Byte (TTFB)](/ttfb/), and the best way to reduce TTFB is to:

- Serve your content as geographically close to your users as possible
- Cache that content so recently-requested content can be served again quickly.

> The best way to do both of these things is to [use a CDN](/content-delivery-networks/). CDNs distribute your resources to [edge servers](https://en.wikipedia.org/wiki/Edge_computing), which are spread across the globe, thus limiting the distance those resources have to travel over the wire to your users. CDNs also usually have fine-grained caching controls that can be customized and optimized for your site's needs.

Reviewer 2: Not sure this link it really right as not just CDNs. Plus edge computing could be confused with edge workers, that you reference later.

```suggestion
The best way to do both of these things is to [use a CDN](/content-delivery-networks/). CDNs distribute your resources to edge servers, which are spread across the globe, thus limiting the distance those resources have to travel over the wire to your users. CDNs also usually have fine-grained caching controls that can be customized and optimized for your site's needs.
```

Author: Done

Many developers are familiar with using a CDN to host static assets, but CDNs can serve and cache HTML documents as well, even those that are dynamically generated.

According to the Web Almanac, only [29%](https://almanac.httparchive.org/en/2022/cdn#cdn-adoption) of HTML document requests were served from a CDN, which means there is significant opportunity for sites to claim additional savings.

Some tips for configuring your CDNs are:

> - Consider increasing how long content is cached for (e.g. is it actually critical that content is always fresh? Or can it be a few minutes stale?).

Author: Done

Despite this, [a significant number of websites](https://almanac.httparchive.org/en/2022/performance#bfcache-eligibility) are ineligible for the bfcache and so are missing out on this free web performance win for a significant number of navigations. Unless your page is loading sensitive information that you don't want to be restored from memory, you'll want to make sure that your pages are eligible.

Site owners should check that their pages are [eligible for the bfcache](/bfcache/#optimize-your-pages-for-bfcache) and work on any reasons why they are not. Chrome already [has a bfcache tester in DevTools](/bfcache/#test-to-ensure-your-pages-are-cacheable) and this year we plan to enhance tooling here with [a new Lighthouse audit performing a similar test](https://github.com/GoogleChrome/lighthouse/issues/13960) and [an API to measure this in the field](https://chromestatus.com/feature/5684908759449600).

> While we have included the bfcache in the CLS section, as we saw the biggest gains there so far, the bfcache will also improve the other Core Web Vitals, i.e. with 0 LCP and much improved FID and INP. It is one of [a number of instant navigations](https://calendar.perfplanet.com/2022/fast-is-good-instant-is-better/) available to drastically improve page navigations.

Reviewer: ```suggestion
While we have included the bfcache in the CLS section, as we saw the biggest gains there so far, the bfcache will generally also improve other Core Web Vitals too. It is one of [a number of instant navigations](https://calendar.perfplanet.com/2022/fast-is-good-instant-is-better/) available to drastically improve page navigations.
```

Good spot, and I think this is actually inaccurate anyway since LCP will not be 0 and I don't think we have any evidence to show that bfcache improves INP.

> While we have included the bfcache in the CLS section, as we saw the biggest gains there so far, the bfcache will also improve the other Core Web Vitals, i.e. with 0 LCP and much improved FID and INP. It is one of [a number of instant navigations](https://calendar.perfplanet.com/2022/fast-is-good-instant-is-better/) available to drastically improve page navigations.

Author: Done

### Avoid animations/transitions that use layout-inducing CSS properties

> Another common source of layout shifts is when elements are animated. For example, cookie banners or other notification banners can push content out of their way, and will cause a layout shift if they do.

Author: Done

> While HTTP Archive data can't conclusively connect animations to layout shifts, the data does show that pages that animate any CSS property that *could* affect layout are 15% less likely to have "good" CLS than pages overall. Some properties are associated with worse CLS than others. For instance, pages that animate `margin` or `border` widths have "poor" CLS at almost twice the rate that pages are assessed as poor overall.

Reviewer 3: ```suggestion
While HTTP Archive data can't conclusively connect animations to layout shifts, the data does show that pages that animate any CSS property that *could* affect layout are 15% less likely to have "good" CLS than pages overall. Some properties are associated with worse CLS than others. For instance, pages that animate `margin` or `border` widths have "poor" CLS at almost twice the rate that pages overall are assessed as poor.
```

Author: Done

> If an element must change its location or the area it takes up, the easiest way to avoid these shifts is to pull the animations outside of the normal document flow. The most common way is to **use animations that don't affect layout**, like `transform`. Animating with `transform` can have the added benefit of [improving page performance overall](/animations-guide/) by moving more animation work onto the GPU and off the main thread. A page could also remove the element itself from the document flow, like with `position: absolute` or `fixed`, though animating some properties (like `top`) could still trigger a layout shift on the element itself.

Reviewer: ```suggestion
This is perhaps not surprising because any time you transition or animate *any* layout-inducing CSS property, it will result in [layout shifts](/cls/#layout-shifts-in-detail), and if those layout shifts are not within 500 milliseconds of a user interaction, they will impact CLS.

What may be surprising to some developers is that this is true even in cases where the element is taken outside of the normal document flow. For example, absolutely positioned elements that animate `top` or `left` will cause layout shifts, even if they aren't pushing other content around. However, if instead of animating `top` or `left` you animate `transform:translateX()` or `transform:translateY()`, it won't cause the browser to update page layout and thus won't produce any layout shifts.

Preferring animation of CSS properties that can be updated on the browser's compositor thread has long been [a performance best practice](/animations-guide/) because it moves that work onto the GPU and off the main thread. And in addition to it being a general performance best practice, it can also help improve CLS.

As a general rule, never animate or transition any CSS property that requires the browser to update the page layout, unless you're doing it in response to a user tap or key press (though [not `hover`](/cls/#user-initiated-layout-shifts)). And whenever possible, prefer transitions and animations using the CSS [`transform`](https://developer.mozilla.org/docs/Web/CSS/transform) property.
```

Author: Done

> The Lighthouse audit [Avoid non-composited animations](https://developer.chrome.com/docs/lighthouse/performance/non-composited-animations/) will warn when a page animates potentially slow CSS properties, a large number of which can cause layout shifts.

Author: Done

Our new [Interaction to Next Paint (INP)](/inp/) metric is a possible successor to FID, and all of the recommendations below apply equally well to both FID and INP. Given that sites [perform worse](https://almanac.httparchive.org/en/2022/performance#inp-as-a-hypothetical-cwv-metric) on INP than FID, especially on mobile, we encourage developers to seriously consider these responsiveness recommendations, despite having "good" FID.

### Avoid or break up long tasks

Tasks are any piece of discrete work that the browser does. Tasks include rendering, layout, parsing, and compiling and executing scripts. When tasks become [long tasks](/long-tasks-devtools/#what-are-long-tasks)—that is, 50 milliseconds or longer—they block the main thread from being able to respond quickly to user inputs.

Per the Web Almanac, there's [plenty of evidence](https://almanac.httparchive.org/en/2022/javascript#long-tasksblocking-time) to suggest that developers could be doing more to avoid or break up long tasks. While breaking up long tasks may not be as low of an effort as other recommendations in this article, it's less effort than other techniques not offered in this article.

> While you should always strive to do as little work as possible in JavaScript, you can help the main thread quite a bit by [breaking up long tasks into smaller ones](/optimize-long-tasks/). You can accomplish this by **[yielding to the main thread](/optimize-long-tasks/#use-asyncawait-to-create-yield-points)** often so that rendering updates and other user interactions can occur more quickly.

Author: Done

> Another option is to consider using APIs such as `isInputPending` and the Scheduler API. [`isInputPending`](/optimize-long-tasks/#yield-only-when-necessary) is a function that returns a boolean value that indicates whether a user input is pending. If it returns true, you can yield to the main thread so it can handle the pending user input. [The Scheduler API](/optimize-long-tasks/#a-dedicated-scheduler-api) is a more advanced approach, which allows you to schedule work based on a system of priorities that take into account whether the work being done is user-visible or backgrounded.

Reviewer 4: ```suggestion
Another option is to consider using APIs such as `isInputPending` and the Scheduler API. [`isInputPending`](/optimize-long-tasks/#yield-only-when-necessary) is a function that returns a boolean value that indicates whether a user input is pending. If it returns `true`, you can yield to the main thread so it can handle the pending user input.

The [Scheduler API](/optimize-long-tasks/#a-dedicated-scheduler-api) is a more advanced approach, which allows you to schedule work based on a system of priorities that take into account whether the work being done is user-visible or backgrounded.
```

Author: Done

> Another option is to consider using APIs such as `isInputPending` and the Scheduler API. [`isInputPending`](/optimize-long-tasks/#yield-only-when-necessary) is a function that returns a boolean value that indicates whether a user input is pending. If it returns true, you can yield to the main thread so it can handle the pending user input. [The Scheduler API](/optimize-long-tasks/#a-dedicated-scheduler-api) is a more advanced approach, which allows you to schedule work based on a system of priorities that take into account whether the work being done is user-visible or backgrounded.

Author: Done

- [Keep your DOM size small](https://developer.chrome.com/docs/lighthouse/performance/dom-size/). DOM size and the intensity of layout work are correlated. When the renderer has to update the layout for a very large DOM, the work required to recalculate its layout can increase significantly.
- [Use CSS containment](https://developer.mozilla.org/docs/Web/CSS/CSS_Containment). CSS containment relies on the CSS `contain` property, which gives instructions to the browser about how to do layout work for the container the `contain` property is set on, including even isolating the scope of layout and rendering to a specific root in the DOM. It's not always an easy process, but by isolating areas containing complex layouts, you can avoid doing layout and rendering work for them that isn't necessary.

## Conclusion

> Improving page performance can seem like a daunting task, especially when you consider that there is a mountain of guidance across the web to consider. By focusing on these recommendations, however, you can approach the problem with focus and purpose, and hopefully move the needle for your website's Core Web Vitals in 2023.

Author: Done

Based on careful analysis of the state of the web, these recommendations represent the subset of best practices that we believe are the *most effective* ways to improve Core Web Vitals performance. If you find that some or all of this guidance doesn't apply to your specific situation, there's always more you can do. Check out these optimization guides for each of the metrics discussed in this article for more information:

- [Optimize LCP](/optimize-lcp/)
- [Optimize CLS](/optimize-cls/)
- [Optimize FID](/optimize-fid/)
- [Optimize INP](/optimize-inp/)

Here's to a new year, and a faster web for all! May your sites be fast for your users in all the ways that matter most.

> 
> 
> _Photo by [Devin Avery](https://unsplash.com/@devintavery)_

Reviewer 4: ```suggestion

_Photo by [Devin Avery](https://unsplash.com/@devintavery)_
```

Author: Done
